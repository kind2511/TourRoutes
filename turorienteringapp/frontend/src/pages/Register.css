/* 
Common styles for the background of both the Register and Sign-in pages.
*/
.register-background,
.signin-background {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    margin-top: -30px; /* here to move the container little bit up */
    justify-content: center; /* Centers content vertically */
    

    
}

/* 
Common styles for the main container of both Register and Sign-in pages.
*/
.register-container,
.signin-container {
    width: 500px;
    padding: 60px 60px 60px 40px; /* Reduced left padding from 60px to 40px */
    border-radius: 10px;
    box-shadow: 0px 0px 15px rgba(51, 240, 4, 0.1);
    background-color: #d5d2d2;
    margin-top:  50px; /*Adjusting the top margin*/
    
}

/* 
Specific styles for the heading (h2) inside the Register and Sign-in containers.
*/
.register-container h2,
.signin-container h2 {
    margin-bottom: 28px;
    text-align: left;
    color: #333;
    font-weight: bold;  /* Bold font-weight for the heading */
    font-size: 30px;
}

/* 
Styles for labels inside .input-group. 
*/
.input-group label {
    display: block;
    margin-bottom: 5px;
    color: #555;
}

/* 
Styles for input fields in both Register and Sign-in containers.
*/
.register-container input,
.signin-container input {
    width: 100%;
    padding: 12px 15px;
    border: 1px solid #e0e0e0;
    border-radius: 5px;
    font-size: 14px;
    transition: border 0.3s;
}

/* 
Styles for focused input fields in both containers.
They have a specific border color and no outline.
*/
.register-container input:focus,
.signin-container input:focus {
    border-color: #4CAF50;
    outline: none;
}

/* 
Styles for buttons in both containers.
*/
.register-container button,
.signin-container button {
    width: 100%;
    padding: 12px;
    border: 1px solid #4CAF50;  /* Green border */
    border-radius: 5px;
    background-color: #4CAF50;
    color: white;
    cursor: pointer;
    text-decoration: none;  /* No underline */
    transition: background 0.3s;
    margin-top: 30px; /*move the button little bit down*/
}

/* 
Hover effect for buttons in both containers. 
On hover, the background color changes.
*/
.register-container button:hover,
.signin-container button:hover {
    background-color: #45a049;
}

/* 
Common styles for the prompt message in both Register and Sign-in pages. 
*/
.login-prompt,
.register-prompt {
    margin-top: 20px;
    text-align: center;
}

/* 
Styles for clickable links in the prompt message. 
*/
.login-link,
.register-link {
    color: #4CAF50;
    cursor: pointer;
    text-decoration: underline;
}

/* 
Styles for show-password and toggle-password elements. 
*/
.show-password,
.toggle-password {
    display: block;
    margin-top: 5px;
    color: #4CAF50;
    cursor: pointer;
    font-size: 12px;
    text-decoration: underline;  
}

/* 
Styles for the forgot-password link. 
*/
.forgot-password {
    text-decoration: underline;
    color: #45a049;  
}

/*
 * Styling for the logo that's shared with the navigation bar.
 */
.nav-logo {
    font-family: 'Pacifico', cursive;
    font-size: 38px;
    font-weight: bold;
    color: #4CAF50;
    letter-spacing: 1px;
    text-transform: capitalize;
    cursor: pointer;
    text-decoration: none;
    display: block;
    text-align: center;
    margin-bottom: 20px;  /* Added to provide some spacing after the logo */
}


.register-container input + input {
    margin-top: 20px;
}

