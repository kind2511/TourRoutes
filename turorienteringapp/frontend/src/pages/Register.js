import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import ReCAPTCHA from "react-google-recaptcha";
import './Register.css';

function Register() {

  // State for form fields
  const [formData, setFormData] = useState({
    email: '',
    firstName: '',
    lastName: '',
    password: '',
    confirmPassword: ''
  });

  // State for reCAPTCHA value
  const [recaptchaValue, setRecaptchaValue] = useState(null);

  // State for error messages
  const [errorMessage, setErrorMessage] = useState("");

  // React Router hook to programmatically navigate
  const navigate = useNavigate();

  // Update form data state on input change
  const handleInputChange = (event) => {
    const { name, value } = event.target;
    setFormData(prevState => ({ ...prevState, [name]: value }));

    // Check password length for the password field on-the-fly
    if (name === "password") {
      if (value.length < 8) {
        setErrorMessage("Password must be at least 8 characters long!");
      } else {
        setErrorMessage("");  // Clear the error message 
      }
    }
  }

  // Handle form submission and user registration
  const handleSubmit = async (e) => {
    e.preventDefault();

    // Check if reCAPTCHA is completed
    if (!recaptchaValue) {
      setErrorMessage("Please confirm that you are not a robot!");
      return;
    }

    if (formData.password.length < 8) {
      setErrorMessage("Password must be at least 8 characters long!")
      return;
    }

    if (formData.password !== formData.confirmPassword) {
      setErrorMessage("Passwords don't match!");
      return;
    }

    try {
      setErrorMessage(""); // Clear any previous errors
      const response = await fetch('http://localhost:8000/api/v1/users/register', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(formData)
      });
      const data = await response.json();

      // Check if the registration was successful
      if (response.ok) {
        // Navigate to the Welcome page after successful registration
        navigate('/welcome');
      } else {
        // Display error message from server or a default message
        setErrorMessage(data.message || "Failed to register.");
      }
    } catch (error) {
      console.error('Error during registration:', error);
      setErrorMessage('Error during registration. Please try again later.');
    }
  }

  // Navigate to login page
  const handleLoginClick = () => {
    navigate('/login');
  }

  return (
    <div className="register-background">
      <div className="register-container">
        {/* Navigate to homepage when the TurRuter logo is clicked */}
        <span className="nav-logo" onClick={() => navigate('/')}>TurRuter</span>

        {errorMessage && <p className="error-message">{errorMessage}</p>} {/* Display error message if it exists */}

        <form onSubmit={handleSubmit}>
          <label htmlFor="firstName">First Name</label>
          <input type="text" name="firstName" id="firstName" value={formData.firstName} onChange={handleInputChange} required autoComplete="given-name" />

          <label htmlFor="lastName">Last Name</label>
          <input type="text" name="lastName" id="lastName" value={formData.lastName} onChange={handleInputChange} required autoComplete="family-name" />

          <label htmlFor="email">Email Address</label>
          <input type="email" name="email" id="email" value={formData.email} onChange={handleInputChange} required autoComplete="email" />

          <label htmlFor="password">Password</label>
          <input type="password" name="password" id="password" value={formData.password} onChange={handleInputChange} required autoComplete="new-password" />

          <label htmlFor="confirmPassword">Confirm Password</label>
          <input type="password" name="confirmPassword" id="confirmPassword" value={formData.confirmPassword} onChange={handleInputChange} required autoComplete="new-password" />

          {/* reCAPTCHA widget */}
          <ReCAPTCHA
            sitekey="6LetW-koAAAAAJsLZtyr9Lf7ShzcAhXntNqU60Ke"
            onChange={value => setRecaptchaValue(value)}
          />

          <button type="submit">Register</button>
        </form>

        <div className="login-prompt">
          Already a member? <span className="login-link" onClick={handleLoginClick}>Log In</span>
        </div>
      </div>
    </div>
  );
}

export default Register;
