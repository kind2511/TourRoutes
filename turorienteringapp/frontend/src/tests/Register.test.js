import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Register from '../pages/Register';

test('Register renders without crashing', () => {
    render(
        <BrowserRouter>
            <Register />
        </BrowserRouter>
    );
});

//---------------------------------------->

// Snapshot test for the Regsiter component
test('Regiser component snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  


