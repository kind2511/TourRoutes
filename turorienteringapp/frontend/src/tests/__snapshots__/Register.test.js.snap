// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regiser component snapshot 1`] = `
<DocumentFragment>
  <div
    class="register-background"
  >
    <div
      class="register-container"
    >
      <span
        class="nav-logo"
      >
        TurRuter
      </span>
       
      <form>
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          autocomplete="given-name"
          id="firstName"
          name="firstName"
          required=""
          type="text"
          value=""
        />
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          autocomplete="family-name"
          id="lastName"
          name="lastName"
          required=""
          type="text"
          value=""
        />
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="new-password"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <label
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <input
          autocomplete="new-password"
          id="confirmPassword"
          name="confirmPassword"
          required=""
          type="password"
          value=""
        />
        <div />
        <button
          type="submit"
        >
          Register
        </button>
      </form>
      <div
        class="login-prompt"
      >
        Already a member? 
        <span
          class="login-link"
        >
          Log In
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
